### PR TITLE
removing background state update of Request object by RequestConverte…

### PR DESCRIPTION
**Pull request by kutty-kumar**
Changes required for removing background state update of request object by RequestConverters.Params object
[https://github.com/elastic/elasticsearch/issues/39666](https://github.com/elastic/elasticsearch/issues/39666)

### CODE REVIEW

**Comment by javanna** [comment]
jenkins test this please

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityRequestConverters.java [hunk 1]
```diff
@@ -153,17 +158,19 @@ static Request hasPrivileges(HasPrivilegesRequest hasPrivilegesRequest) throws I
...
+        if (!clearRealmCacheRequest.getRealms().isEmpty()) {
             builder.addCommaSeparatedPathParts(clearRealmCacheRequest.getRealms().toArray(Strings.EMPTY_ARRAY));
         } else {
             builder.addPathPart("_all");
         }
         final String endpoint = builder.addPathPartAsIs("_clear_cache").build();
         Request request = new Request(HttpPost.METHOD_NAME, endpoint);
-        if (clearRealmCacheRequest.getUsernames().isEmpty() == false) {
-            RequestConverters.Params params = new RequestConverters.Params(request);
+        if (!clearRealmCacheRequest.getUsernames().isEmpty()) {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityRequestConverters.java]: can you please revert this change and all the similar ones? we tend to prefer the `boolean == false` over `!boolean` as the exclamation mark can easily be missed.

File: client/rest/src/main/java/org/elasticsearch/client/Request.java [hunk 2]
```diff
@@ -50,6 +49,14 @@ public Request(String method, String endpoint) {
         this.endpoint = Objects.requireNonNull(endpoint, "endpoint cannot be null");
     }
 
+    /**
+     * Copy the parameters from {@literal RequestConverters.Params} into {@linkplain Request}.
```
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/Request.java]: I think that we should just say "Set the parameters" , who sets them may change in the future

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java [hunk 3]
```diff
@@ -756,15 +779,19 @@ public static ContentType createContentType(final XContentType xContentType) {
      * a {@link Request} and adds the parameters to it directly.
      */
     static class Params {
-        private final Request request;
+        private final Map<String, String> requestParamsMap;
+
+        Params() {
+            this.requestParamsMap = new HashMap<>();
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: nit: you could initialize the map above and remove the default constructor.

File: client/rest/src/main/java/org/elasticsearch/client/Request.java (hunk 2 above)
Comment (javanna) [reply client/rest/src/main/java/org/elasticsearch/client/Request.java]: Also, note that RequestConverters are part of the high-level REST client while Request is part of the low-level REST client which is used independently.

File: client/rest/src/main/java/org/elasticsearch/client/Request.java [hunk 4]
```diff
@@ -50,6 +49,14 @@ public Request(String method, String endpoint) {
...
+     * @param requestConverterParams the request converter parameter map
```
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/Request.java]: can we remove the mention of request converters here?

File: client/rest/src/main/java/org/elasticsearch/client/Request.java [hunk 5]
```diff
@@ -50,6 +49,14 @@ public Request(String method, String endpoint) {
...
+     */
+    public void setParameters(Map<String, String> requestConverterParams){
```
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/Request.java]: and rename the argument to not refer to request converters?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java [hunk 6]
```diff
@@ -756,15 +779,19 @@ public static ContentType createContentType(final XContentType xContentType) {
...
-        Params(Request request) {
-            this.request = request;
+        Map<String, String> getRequestParams(){
+            return this.requestParamsMap;
         }
 
         Params putParam(String name, String value) {
             if (Strings.hasLength(value)) {
-                request.addParameter(name, value);
+                requestParamsMap.put(name, value);
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: with this change we lose two checks that we were previously performing by calling Request#addParameter: 
- check that the key is not null
- check that the param has not already been set (it would get overridden by default, but we prefer to throw error)

I think that we would want to keep these two checks.

**Comment by javanna** [comment]
thanks @kutty-kumar for opening this PR! I just did a first quick review and left some comments, also I started tests which are running. Let me know if you have any question on addressing the comments.

**Comment by javanna** [comment]
@kutty-kumar it seems like there's a couple of failing tests. It's fine not to run all of our tests before submitting a PR, but in your case, could you run `../../gradlew check` from `client/rest-high-level` directory please? that should help you reproduce the problem. Let me know if you need help fixing those failures.

**Comment by kutty-kumar** [comment]
@javanna thanks for the initial review, i will work on resolving them and update the PR.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java (hunk 6 above)
Comment (kutty-kumar) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: @javanna what error should i throw? i think that check would move to the copy method of Request class
```suggestion
                   public void setParameters(Map<String, String> paramSource){
        paramSource.keySet().forEach(key -> {
            if(this.parameters.get(key) != null){
                throw new Error(String.format("%s is already set", key));
            }
        });
        this.parameters.putAll(paramSource);
    }
```
does this look good?

Comment (javanna) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: can you throw the same error that we throw in the Request class please? IllegalArgumentException?

**Comment by kutty-kumar** [comment]
@javanna i have made the required changes please review them also i am currently fixing the tests, will update here once done. Thanks

File: client/rest/src/main/java/org/elasticsearch/client/Request.java [hunk 7]
```diff
@@ -50,6 +49,19 @@ public Request(String method, String endpoint) {
...
+     * copy the params into {@linkplain Request}.
+     * @param paramSource map holding the properties for request to copy from
+     */
+    public void setParameters(Map<String, String> paramSource){
+        paramSource.keySet().forEach(key -> {
+            if(this.parameters.get(key) != null){
+                throw new IllegalArgumentException(String.format("%s is already set", key));
+            }
+        });
+        this.parameters.putAll(paramSource);
```
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/Request.java]: should this method add parameters to the existing ones or just set them all at once? You could add parameters through `addParameter` and currently we would keep on adding through this new method. But that is not what its name `setParameters` suggests. Shall we rename this to `addParameters` ?

File: client/rest/src/main/java/org/elasticsearch/client/Request.java [hunk 8]
```diff
@@ -150,7 +162,7 @@ public String toString() {
         b.append("Request{");
         b.append("method='").append(method).append('\'');
         b.append(", endpoint='").append(endpoint).append('\'');
-        if (false == parameters.isEmpty()) {
+        if (!parameters.isEmpty()) {
```
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/Request.java]: please revert this change

File: client/rest-high-level/src/main/java/org/elasticsearch/client/WatcherRequestConverters.java [hunk 9]
```diff
@@ -69,15 +69,16 @@ static Request putWatch(PutWatchRequest putWatchRequest) {
             .build();
 
         Request request = new Request(HttpPut.METHOD_NAME, endpoint);
-        RequestConverters.Params params = new RequestConverters.Params(request)
+        RequestConverters.Params params = new RequestConverters.Params()
             .withIfSeqNo(putWatchRequest.ifSeqNo())
             .withIfPrimaryTerm(putWatchRequest.ifPrimaryTerm());
-        if (putWatchRequest.isActive() == false) {
+        if (!putWatchRequest.isActive()) {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/WatcherRequestConverters.java]: please revert

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityRequestConverters.java (hunk 1 above)
Comment (javanna) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityRequestConverters.java]: can you please revert this?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityRequestConverters.java [hunk 10]
```diff
@@ -153,17 +158,19 @@ static Request hasPrivileges(HasPrivilegesRequest hasPrivilegesRequest) throws I
     static Request clearRealmCache(ClearRealmCacheRequest clearRealmCacheRequest) {
         RequestConverters.EndpointBuilder builder = new RequestConverters.EndpointBuilder()
             .addPathPartAsIs("_security/realm");
-        if (clearRealmCacheRequest.getRealms().isEmpty() == false) {
+        if (!clearRealmCacheRequest.getRealms().isEmpty()) {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityRequestConverters.java]: revert this?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java [hunk 11]
```diff
@@ -818,7 +841,7 @@ Params withPreference(String preference) {
         }
 
         Params withRealtime(boolean realtime) {
-            if (realtime == false) {
+            if (!realtime) {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: revert

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java [hunk 12]
```diff
@@ -785,7 +808,7 @@ Params withDocAsUpsert(boolean docAsUpsert) {
 
         Params withFetchSourceContext(FetchSourceContext fetchSourceContext) {
             if (fetchSourceContext != null) {
-                if (fetchSourceContext.fetchSource() == false) {
+                if (!fetchSourceContext.fetchSource()) {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: revert

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java [hunk 13]
```diff
@@ -926,7 +949,7 @@ Params withIndicesOptions(IndicesOptions indicesOptions) {
                 withIgnoreUnavailable(indicesOptions.ignoreUnavailable());
                 putParam("allow_no_indices", Boolean.toString(indicesOptions.allowNoIndices()));
                 String expandWildcards;
-                if (indicesOptions.expandWildcardsOpen() == false && indicesOptions.expandWildcardsClosed() == false) {
+                if (!indicesOptions.expandWildcardsOpen() && !indicesOptions.expandWildcardsClosed()) {
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: revert

File: client/rest/src/main/java/org/elasticsearch/client/Request.java (hunk 7 above)
Comment (javanna) [reply client/rest/src/main/java/org/elasticsearch/client/Request.java]: If we rename to addParameters we could just make this method call addParameter for each entry in the map, then there is also no duplication around null key check and duplicate check. Does this make sense?

**Review by javanna — CHANGES_REQUESTED** [review]
I left some more comments, let me know if you need help with the test failures.

**Comment by kutty-kumar** [comment]
@javanna yes i need some help in fixing the tests, not able to figure out where exactly they are failing, please direct me towards that, i will take care of the changes
Thanks

**Comment by javanna** [comment]
retest this please

**Comment by kutty-kumar** [comment]
jenkins test this please

**Comment by javanna** [comment]
@kutty-kumar the magic jenkins commands work only if you are part of the Elastic organization, for security reasons, I will run the tests again.

**Comment by javanna** [comment]
retest this please

**Comment by kutty-kumar** [comment]
@javanna i have started the gradle check process, its still running, doesn't seem to have any failures. Hence i wanted to test it on jenkins, thanks for the info :)

**Comment by javanna** [comment]
@kutty-kumar there are still two failures:
```
org.elasticsearch.client.IndicesRequestConvertersTests.testExistsAlias
org.elasticsearch.client.SecurityRequestConvertersTests.testDeleteUser
```

would you mind having a look at those? You can run client tests only by executing `../../gradlew check`  from `client/rest-high-level` folder. Thanks

**Comment by kutty-kumar** [comment]
done @javanna, please rebuild the source.
Thanks

**Comment by javanna** [comment]
retest this please

**Comment by kutty-kumar** [comment]
`➜  rest-high-level git:(refactoring/requestConverterParams) ✗ ../../gradlew check

> Configure project :benchmarks
=======================================
Elasticsearch Build Hamster says Hello!
  Gradle Version        : 5.2.1
  OS Info               : Mac OS X 10.14.3 (x86_64)
  JDK Version           : 11 (Oracle Corporation 11.0.2 [Java HotSpot(TM) 64-Bit Server VM 11.0.2+9-LTS])
  JAVA_HOME             : /Library/Java/JavaVirtualMachines/jdk-11.0.2.jdk/Contents/Home
  Random Testing Seed   : 5FC48F357ECA8590
=======================================

> Task :client:rest-high-level:unitTest
==> Test Info: seed=5FC48F357ECA8590; jvms=4; suites=275

[ant:junit4] JVM J0: stderr was not empty, see: /Users/halodoc/Projects/elasticsearch1/client/rest-high-level/build/testrun/unitTest/temp/junit4-J0-20190326_162010_90016372490423214617704.syserr
[ant:junit4] JVM J2: stderr was not empty, see: /Users/halodoc/Projects/elasticsearch1/client/rest-high-level/build/testrun/unitTest/temp/junit4-J2-20190326_162010_90018010291334947152229.syserr
[ant:junit4] JVM J3: stderr was not empty, see: /Users/halodoc/Projects/elasticsearch1/client/rest-high-level/build/testrun/unitTest/temp/junit4-J3-20190326_162010_9011526902967533266904.syserr

> Task :client:rest-high-level:unitTest
==> Test Summary: 275 suites, 870 tests

> Task :client:rest-high-level:integTestRunner
==> Test Info: seed=5FC48F357ECA8590; jvm=1; suites=45

HEARTBEAT J0 PID(12684@Kumards-MacBook-Pro.local): 2019-03-26T16:27:33, stalled for 10.7s at: SecurityDocumentationIT.testInvalidateToken

> Task :client:rest-high-level:integTestRunner
Slow Tests Summary:
102.99s | org.elasticsearch.client.SearchIT
 93.71s | org.elasticsearch.client.MachineLearningIT
 89.45s | org.elasticsearch.client.documentation.MlClientDocumentationIT
 50.11s | org.elasticsearch.client.documentation.SecurityDocumentationIT
 45.63s | org.elasticsearch.client.IndicesClientIT

==> Test Summary: 45 suites, 490 tests

Deprecated Gradle features were used in this build, making it incompatible with Gradle 6.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/5.2.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 13m 16s
273 actionable tasks: 56 executed, 217 up-to-date`
Did that @javanna

**Comment by javanna** [comment]
can you please merge master in @kutty-kumar ?

**Comment by kutty-kumar** [comment]
@javanna its already up to date with master, please let me know what exactly i should do.
Thanks

**Comment by javanna** [comment]
retest this please

**Comment by javanna** [comment]
@kutty-kumar it does not look up-to-date to me, could you pull latest master (doing `git checkout master`, then `git pull origin master`) and then merge it in your branch (by doing `git checkout refactoring/requestConverterParams`, then `git merge master`) and then push your updated branch to your remote?

**Comment by kutty-kumar** [comment]
@javanna can you please check now, my fork was behind the actual repository, hence the branch was behind, i have taken the changes from the actual repository and force pushed.

Thanks

**Comment by javanna** [comment]
retest this please

**Comment by javanna** [comment]
@kutty-kumar some code needs updating now, as there are more users of the internal API that you have modified in your change. It should be straight-forward to address.

**Comment by javanna** [comment]
hi @kutty-kumar have you had a chance to look into the needed changes? Do you need help with this?

**Comment by kutty-kumar** [comment]
Hi @javanna i thought that i had to refactor a lot, hence i wanted your suggestions on doing this the right way. Please add in some thoughts which i could use to refine the code.

**Comment by javanna** [comment]
@kutty-kumar as far as I can see, while you PR was getting reviewed, new API were added to the high-level REST client which use the mechanism that you were refactoring. You need to make the same changes to those newly added API.

**Comment by jasontedor** [comment]
@kutty-kumar Are you interested in bringing this PR to completion?

**Comment by kutty-kumar** [comment]
@jasontedor  yes i will complete the PR by this weekend, sorry for the delay!

**Comment by javanna** [comment]
This has stalled as no work has been done on it for a couple of months. Feel free to reopen or open a new PR when/if you get the time to work on it again.
